src/main/java/io/odpf/depot/http/request/builder/UriBuilder.java

```diff
@@ -8,15 +12,27 @@
 
 public class UriBuilder {
 
-    private final String urlConfig;
+    private final Template urlTemplate;
 
     public UriBuilder(String urlConfig) {
-        this.urlConfig = urlConfig;
+        this.urlTemplate = new Template(urlConfig);
     }
 
-    public URI build(Map<String, String> queryParam) throws URISyntaxException {
-        URIBuilder uriBuilder = new URIBuilder(urlConfig);
-        queryParam.forEach(uriBuilder::addParameter);
-        return uriBuilder.build();
+    public URI build(ParsedOdpfMessage message, OdpfMessageSchema schema, Map<String, String> queryParam) {
+        return build(urlTemplate.parse(message, schema), queryParam);
+    }
+
+    public URI build(Map<String, String> queryParam) {
+        return build(urlTemplate.parse(null, null), queryParam);
```

Comment:
let's add a toString() method on Template which returns the underlying string. 
Passing null is weird here.

Reply:
public URI build(Map<String, String> queryParam) {
        return build(urlTemplate.toString(), queryParam);
    }